diff --git a/include/util.h b/include/util.h
index f46e452..3c2d6ae 100644
--- a/include/util.h
+++ b/include/util.h
@@ -20,6 +20,24 @@
 #include <usual/string.h>
 #include <usual/cfparser.h>
 
+/*
+ * sets the global writer to NULL for the pool
+ */
+void clear_global_writer(PgPool *pool);
+
+/*
+ * get the global writer, if any from the pool
+ */
+PgPool *get_global_writer(PgPool *pool);
+
+/*
+ * checks if fast switchover mode is disabled for the specific pool.
+ * Fast switchover is enabled only when:
+ * 1. Pool has topology query configured and
+ * 2. fast switchover flag is enabled
+ */
+bool is_fast_switchover_disabled_for_pool(PgPool *pool);
+
 /*
  * logging about specific socket
  */
