--- a/src/client.c	2024-10-14 15:33:21
+++ b/src/client.c	2024-10-14 15:33:20
@@ -1106,7 +1106,7 @@ static bool handle_client_startup(PgSocket *client, Pk
 			return false;
 		}
 
-		if (client->pool && !client->wait_for_user_conn && !client->wait_for_user) {
+		if (client->pool && !client->wait_for_user_conn && !client->wait_for_user && !get_global_writer(client->pool)) {
 			disconnect_client(client, true, "client re-sent startup pkt");
 			return false;
 		}
@@ -1232,6 +1232,7 @@ static bool handle_client_work(PgSocket *client, PktHd
 {
 	SBuf *sbuf = &client->sbuf;
 	int track_outstanding = false;
+    int in_transaction;
 	PreparedStatementAction ps_action = PS_IGNORE;
 	PgClosePacket close_packet;
 
@@ -1396,9 +1397,12 @@ static bool handle_client_work(PgSocket *client, PktHd
 		client->pool->stats.query_count++;
 		client->query_start = get_cached_time();
 	}
+
+	/* xact_start is non-zero if we are inside a transaction */
+	in_transaction = client->xact_start != 0;
 
 	/* remember timestamp of the first query in a transaction */
-	if (!client->xact_start) {
+	if (!in_transaction) {
 		client->pool->stats.xact_count++;
 		client->xact_start = client->query_start;
 	}
@@ -1406,6 +1410,15 @@ static bool handle_client_work(PgSocket *client, PktHd
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
 
+	/* pgbouncer-rr extensions: query rewrite & client connection routing */
+	if (pkt->type == 'Q' || pkt->type == 'P') {
+		if (!rewrite_query(client, in_transaction, pkt)) {
+			return false;
+		}
+		route_client_connection(client, in_transaction, pkt);
+	}
+
+
 	/* acquire server */
 	if (!find_server(client))
 		return false;
