diff --git a/src/util.c b/src/util.c
index fad0172..eaaadc0 100644
--- a/src/util.c
+++ b/src/util.c
@@ -32,6 +32,33 @@
 #include <openssl/evp.h>
 #endif
 
+PgPool *get_global_writer(PgPool *pool)
+{
+	if (!pool)
+		return NULL;
+
+	if (pool->global_writer)
+		return pool->global_writer;
+
+	if (!pool->parent_pool)
+		return NULL;
+
+	return pool->parent_pool->global_writer;
+}
+
+void clear_global_writer(PgPool *pool)
+{
+	if (pool->global_writer) {
+		pool->global_writer = NULL;
+		return;
+	}
+
+	if (!pool->parent_pool)
+		return;
+
+	pool->parent_pool->global_writer = NULL;
+}
+
 int log_socket_prefix(enum LogLevel lev, void *ctx, char *dst, unsigned int dstlen)
 {
 	const struct PgSocket *sock = ctx;
@@ -532,3 +559,7 @@ bool check_reserved_database(const char *value)
 	}
 	return true;
 }
+
+bool is_fast_switchover_disabled_for_pool(PgPool *pool) {
+	return !pool->db->topology_query || !fast_switchover;
+}
